### PR TITLE
New serverless pattern - AWS Step Functions to Amazon Bedrock Integration (SAM)

### PR DESCRIPTION
**Pull request by biswanathmukherjee**
*Issue #, if available:*

*Description of changes:*
The SAM template deploys a Step Functions Express workflow that invokes Amazon Bedrock Anthropic Claud Model v2.1 and returns the generated contents based on the provided prompts. The SAM template contains the required resouces with IAM permission to run the application with logging enabled.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: stepfunctions-bedrock-sam/template.yaml [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+                  - "logs:DescribeLogGroups"
+                Resource: "*"         
+        - PolicyName: BedrockAccess
+          PolicyDocument:
+            Version: "2012-10-17"
+            Statement:
+              - Effect: Allow
+                Action:
+                  - "bedrock:InvokeModel"
+                Resource: "*"
```
Comment (ellisms) [inline stepfunctions-bedrock-sam/template.yaml:75]: Add the Claude ARN to the resource list for least privilege access.

**Review by ellisms — CHANGES_REQUESTED** [review]
Tested and this works. Just a few minor requested changes for you to review.

File: stepfunctions-bedrock-sam/template.yaml (hunk 1 above)
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/template.yaml]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+![End to End Architecture](image/architecture.png)
+
+
+## Testing
+
+Run the following AWS CLI command to send a 'start-sync-execution` comand to start the Step Functions workflow. Note, you must edit the {StateMachineExpressSyncToBedrockArn} placeholder with the ARN of the deployed Step Functions workflow. This is provided in the stack outputs.
+
+```bash
+aws stepfunctions start-sync-execution  --name "test" --state-machine-arn "{StateMachineExpressSyncToBedrockArn}" --input "{\"prompt\": \"\n\nHuman:Write 5 lines about how Moon was formed.\n\nAssistant:\"}" --region {your-region}
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md:60]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+* If the integration fails, the Step Functions workflow will retry up to 5 times before exiting with a `status:FAILED` response.
+
+Please refer to the architecture diagram below:
+
+![End to End Architecture](image/architecture.png)
+
+
+## Testing
+
+Run the following AWS CLI command to send a 'start-sync-execution` comand to start the Step Functions workflow. Note, you must edit the {StateMachineExpressSyncToBedrockArn} placeholder with the ARN of the deployed Step Functions workflow. This is provided in the stack outputs.
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    Once you have run `sam deploy --guided` mode once and saved arguments to a configuration file (samconfig.toml), you can use `sam deploy` in future to use these defaults.
+
+5. Note the outputs from the SAM deployment process. These contain the resource names and/or ARNs which are used for testing.
+
+## How it works
+
+* Start the Express Workflow using the `start-sync-execution` api command with a "prompt" string in English for generating content as per the input payload.
+* The Express Workflow invokes Amazon Bedrock Anthropic Claud v2.1 Model with the prompt.
+* Amazon Bedrock returns the generated content as the given promptsentiment of the input text. 
+* If the integration works fine, the generated content is returned in Step Function execution results within a `output` object.
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    * Allow SAM CLI to create IAM roles with the required permissions.
+
+    Once you have run `sam deploy --guided` mode once and saved arguments to a configuration file (samconfig.toml), you can use `sam deploy` in future to use these defaults.
+
+5. Note the outputs from the SAM deployment process. These contain the resource names and/or ARNs which are used for testing.
+
+## How it works
+
+* Start the Express Workflow using the `start-sync-execution` api command with a "prompt" string in English for generating content as per the input payload.
+* The Express Workflow invokes Amazon Bedrock Anthropic Claud v2.1 Model with the prompt.
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+Important: this application uses various AWS services and there are costs associated with these services after the Free Tier usage - please see the [AWS Pricing page](https://aws.amazon.com/pricing/) for details. You are responsible for any AWS costs incurred. No warranty is implied in this example.
+
+## Requirements
+
+* [Create an AWS account](https://portal.aws.amazon.com/gp/aws/developer/registration/index.html) if you do not already have one and log in. The IAM user that you use must have sufficient permissions to make necessary AWS service calls and manage AWS resources.
+* [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
+* [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+* [AWS Serverless Application Model](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-cli-install.html) (AWS SAM) installed
+* You must have enabled the Anthropic Claud Model access on Amazon Bedrock console.
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
+# AWS Step Functions Express Workflow to Amazon Bedrock Anthropic Claud Model v2.1 for Content Generation  
+
+The Step Functions Express Workflow can be started using the AWS CLI or from another service (e.g. API Gateway) to run an express workflow and return the result.
+
+The SAM template deploys a Step Functions Express workflow that invokes Amazon Bedrock Anthropic Claud Model v2.1 and returns the generated contents based on the provided prompts. The SAM template contains the required resouces with IAM permission to run the application with logging enabled.
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
+# AWS Step Functions Express Workflow to Amazon Bedrock Anthropic Claud Model v2.1 for Content Generation  
+
+The Step Functions Express Workflow can be started using the AWS CLI or from another service (e.g. API Gateway) to run an express workflow and return the result.
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md]: Fixed as suggested.

File: stepfunctions-bedrock-sam/README.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
+# AWS Step Functions Express Workflow to Amazon Bedrock Anthropic Claud Model v2.1 for Content Generation
```
[the remark this answers is not in the document]
Comment (biswanathmukherjee) [reply stepfunctions-bedrock-sam/README.md]: Fixed as suggested.

**Comment by ellisms** [comment]
@biswanathmukherjee Thanks. This looks good, I'll merge and put it in the the pipeline for publishing.

**Review by ellisms — APPROVED** [review]

**Comment by boyney123** [comment]
Your pattern is now live! [https://serverlessland.com/patterns/stepfunctions-bedrock-sam](https://serverlessland.com/patterns/stepfunctions-bedrock-sam)

**Pull request merged**